File: lab/javascript/tecnicas/porcentagens.js
// -------------------- Funções principais --------------------

function percent(value, percent) {
  return value * percent / 100;
}

function percent2(value, percent = 0.5) { // 0.5 = 50%
  return value * percent;
}

function discount(value, discount) {
  return value - (value * discount / 100);
}

function discount_decimal(value, discount = 0.5) {
  return value - (value * discount);
}

function discount_factor(value, discount) {
  return value * (1 - discount / 100);
}

function discount_chain(value, ...discounts) {
  for (let d of discounts) {
    value *= (1 - d / 100);
  }
  return value;
}

function increase(value, percent) {
  return value + (value * percent / 100);
}

function increase_decimal(value, discount = 0.5) {
  return value + (value * discount);
}

function increase_factor(value, percent) {
  return value * (1 + percent / 100);
}

function increase_chain(value, ...percents) {
  for (let p of percents) {
    value *= (1 + p / 100);
  }
  return value;
}

// -------------------- Testes --------------------

console.log(percent(200, 15));            // 30
console.log(percent2(200, 0.15));         // 30
console.log(discount(250, 20));           // 200
console.log(discount_factor(250, 20));    // 200
console.log(discount_decimal(250, 0.20)); // 200
console.log(increase(200, 10));           // 220
console.log(increase_factor(200, 10));    // 220
console.log(increase_decimal(200, 0.10)); // 220
console.log(discount_chain(100, 20, 10)); // 72
console.log(increase_chain(100, 10, 20)); // 132

File: lab/c/tecnicas/porcentagens.c
#include <stdio.h>

// -------------------- Funções principais --------------------

double percent(double value, double percent) {
    return value * percent / 100.0;
}

double percent2(double value, double percent) { // 0.5 = 50%
    return value * percent;
}

double discount(double value, double discount) {
    return value - (value * discount / 100.0);
}

double discount_decimal(double value, double discount) {
    return value - (value * discount);
}

double discount_factor(double value, double discount) {
    return value * (1 - discount / 100.0);
}

double discount_chain(double value, int qtd, double discounts[]) {
    for (int i = 0; i < qtd; i++) {
        value *= (1 - discounts[i] / 100.0);
    }
    return value;
}

double increase(double value, double percent) {
    return value + (value * percent / 100.0);
}

double increase_decimal(double value, double discount) {
    return value + (value * discount);
}

double increase_factor(double value, double percent) {
    return value * (1 + percent / 100.0);
}

double increase_chain(double value, int qtd, double percents[]) {
    for (int i = 0; i < qtd; i++) {
        value *= (1 + percents[i] / 100.0);
    }
    return value;
}

// -------------------- Testes --------------------

int main(void) {
    printf("%.2f\n", percent(200, 15));
    printf("%.2f\n", percent2(200, 0.15));
    printf("%.2f\n", discount(250, 20));
    printf("%.2f\n", discount_factor(250, 20));
    printf("%.2f\n", discount_decimal(250, 0.20));
    printf("%.2f\n", increase(200, 10));
    printf("%.2f\n", increase_factor(200, 10));
    printf("%.2f\n", increase_decimal(200, 0.10));

    double descontos[] = {20, 10};
    printf("%.2f\n", discount_chain(100, 2, descontos));

    double aumentos[] = {10, 20};
    printf("%.2f\n", increase_chain(100, 2, aumentos));

    return 0;
}
